### PR TITLE
Tidy Democracy

### PR DESCRIPTION
**Pull request by shawntabrizi**
This is a trivial cleanup of some low hanging fruit in democracy pallet. Mostly ensuring safe math is used when the block number gets REALLY big.

### CODE REVIEW

File: frame/democracy/src/vote.rs [hunk 1]
```diff
@@ -128,6 +128,10 @@ impl<BlockNumber: Ord + Copy + Zero, Balance: Ord + Copy + Zero> PriorLock<Block
 			self.1 = Zero::zero();
 		}
 	}
+
+	pub fn new() -> Self {
```
Comment (kianenigma) [inline frame/democracy/src/vote.rs:132]: this is equal to `Default` which already exists.

File: frame/democracy/src/lib.rs [hunk 2]
```diff
@@ -1513,9 +1524,16 @@ impl<T: Config> Pallet<T> {
...
+				Voting::Delegating {
+					balance, target, conviction, delegations, mut prior, ..
+				} => {
 					// remove any delegation votes to our current target.
 					Self::reduce_upstream_delegation(&target, conviction.votes(balance));
+					let now = frame_system::Pallet::<T>::block_number();
+					let lock_periods = conviction.lock_periods().into();
+					let unlock_block = now
+						.saturating_add(T::VoteLockingPeriod::get().saturating_mul(lock_periods));
+					prior.accumulate(unlock_block, balance);
```
Comment (kianenigma) [inline frame/democracy/src/lib.rs:1536]: this was a bug before right? we updated the delegation but did not update the lock at all.

**Review by kianenigma — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by xlc** [comment]
Why not using the new Defensive trait?

**Comment by kianenigma** [comment]
> Why not using the new Defensive trait?

I actually considered suggesting this, but while practically it has some benefits (it logs error if anything saturates), that's not how defensive traits are meant to work. A defensive operation is one where there is a proof that a condition can never happen, and it ever does, then we want to be defensive and continue. For example: 

```rust
let _ = ensure_can_withdraw()?;
let remainder = withdraw();
remainder.defensive_is_zero(); 
```

has a reason why `remainder` is zero.

There exists no such proof for block numbers never reaching very large values that can cause saturation.

**Comment by xlc** [comment]
I see. I didn't check it closely but I think saturation the block number means a lot of assumptions will not hold anymore when the block number is big enough and therefore resulting unexpected behaviour and the defensive trait will help a bit on notifying that something isn't right.

Take an example:

```
let now = BlockNumber::MAX;
let vote_end_block = now.saturated_add(VotingPeriod::get()); // we will now believe vote_end_block is a future block, which may not be the case if now is large
some_call_assumes_vote_end_block_is_in_future(vote_end_block);
```

So we don't really support now with very big block number, and there isn't much reason to support it, then we can use defensive trait to indicate it is not supported. "This should never fail, if it failed, it means we are in an unexpected/unsupported state."

**Comment by shawntabrizi** [comment]
I think actually saturating here is the right thing to do.

If we overflowed, proposals would be triggered to end in the past or immediately (depending on the logic).

If we checked add, it could be we are locked out of pushing upgrades to the chain since we can't schedule anything which passes the check.

So saturating here makes sense to me as the furthest possible point in the future we can schedule something.

And yes, if we ever get close to a big block number, 100 years or so, probably it will need to be addressed across the board.

**Pull request merged**